### PR TITLE
Feat/export dataset for research

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
source: https://github.com/open-spaced-repetition/srs-benchmark/issues/29#issuecomment-2419282021

### CODE REVIEW

File: proto/anki/scheduler.proto [hunk 1]
```diff
@@ -67,6 +67,7 @@ service BackendSchedulerService {
       returns (ComputeFsrsWeightsResponse);
   // Generates parameters used for FSRS's scheduler benchmarks.
   rpc FsrsBenchmark(FsrsBenchmarkRequest) returns (FsrsBenchmarkResponse);
+  rpc ExportDataset(ExportDatasetRequest) returns (generic.Empty);
```
Comment (dae) [inline proto/anki/scheduler.proto:70]:
```suggestion
    // Used for exporting revlogs for algorithm research.
  rpc ExportDataset(ExportDatasetRequest) returns (generic.Empty);
```

File: pylib/anki/collection.py [hunk 2]
```diff
@@ -420,6 +420,13 @@ def import_json_file(self, path: str) -> ImportLogWithChanges:
     def import_json_string(self, json: str) -> ImportLogWithChanges:
         return self._backend.import_json_string(json)
 
+    def export_dataset_for_research(
+        self, min_entries: int = 0, target_path: str | None = None
+    ) -> None:
+        if target_path is None:
+            target_path = os.path.join(os.path.dirname(self.path), "dataset.pb")
```
Comment (dae) [inline pylib/anki/collection.py:427]: I'd prefer target_path was required instead of falling back to dataset.pb. Otherwise looks good :+1:

Comment (L-M-Sherlock) [reply pylib/anki/collection.py]: Thanks for the advice. I update it.

File: rslib/src/scheduler/fsrs/weights.rs [hunk 3]
```diff
@@ -127,22 +128,46 @@ impl Collection {
...
+            .collect_vec();
+        let cards = self.storage.get_all_card_entries()?;
+        let decks = self
+            .storage
+            .get_all_decks()?
+            .into_iter()
+            .filter_map(|deck| {
+                let parent_id = self
+                    .storage
+                    .parent_decks(&deck)
```
[the remark this answers is not in the document]
Comment (L-M-Sherlock) [reply rslib/src/scheduler/fsrs/weights.rs:148]: I cannot find an existing routine. What about that:

```rust
    pub(crate) fn immediate_parent_deck(&self, child: &Deck) -> Result<Option<Deck>> {
        if let Some(parent_name) = immediate_parent_name(child.name.as_native_str()) {
            if let Some(parent_did) = self.get_deck_id(parent_name)? {
                return self.get_deck(parent_did);
            }
        }
        Ok(None)
    }
```

[the remark this answers is not in the document]
Comment (L-M-Sherlock) [reply rslib/src/scheduler/fsrs/weights.rs:148]: There is a function named `deck_tree`:

https://github.com/ankitects/anki/blob/c1a2b03871058067cda13681c87b24979f484913/rslib/src/decks/tree.rs#L252-L261

Do you mean this one?

[the remark this answers is not in the document]
Comment (dae) [reply rslib/src/scheduler/fsrs/weights.rs:148]: I think we'd need immediate_parent_deck() and then looking up the parent in a hashset of all deck names. Alternatively, you could use deck_tree() and then walk the tree once to determine the parent of each deck. But this is code that will run infrequently and the delay won't be huge, so I'm ok with not making this change.

[the remark this answers is not in the document]
Comment (dae) [reply rslib/src/scheduler/fsrs/weights.rs]: Sorry, I meant immediate_parent_name() - glad to see you figured that out :)

[the remark this answers is not in the document]
Comment (L-M-Sherlock) [reply rslib/src/scheduler/fsrs/weights.rs]: I refactor the code. Now it looks up the parent from the hashmap.

https://github.com/ankitects/anki/pull/3511/commits/469f2861b1028c4b2b7b152978222f6a9309b550

**Comment by dae** [comment]
Thanks :-) It may take a couple of weeks for me to get this onto AnkiWeb - I still need to catch up with the recent FSRS changes as well. Will let you know.

**Comment by dae** [comment]
Finally got updates working again. Here's a small test export. Please confirm it has everything you want, then I can do a bigger run.

https://drive.google.com/file/d/1JmUvu6SELAy9jSXejksjfTBTKmYEQ3mI/view?usp=sharing

Is 20k a good number, or do you want more or less?

**Comment by L-M-Sherlock** [comment]
> Is 20k a good number, or do you want more or less?

20k feels too much. My device will run out of its disk. 10k is enough. Thanks! I will check the test export and give you feedback later.

**Comment by L-M-Sherlock** [comment]
> Please confirm it has everything you want, then I can do a bigger run.

Checked! It's really what I want.

**Comment by L-M-Sherlock** [comment]
Is it possible to restrict the limit to exclude users who have thousands of revlog entries generated by rescheduling on change?

**Comment by dae** [comment]
(sorry, just another post in case you didn't see my edit)

**Comment by Expertium** [comment]
@dae any chance we will get the new dataset before the release of Anki 24.11?

**Comment by L-M-Sherlock** [comment]
@Expertium, I have received the dataset. The benchmark will be updated in next two weeks.

**Pull request merged**